llama_index/query_engine/citation_query_engine.py

```diff
@@ -0,0 +1,227 @@
+    """Citation query engine."""
+
+    @classmethod
+    def from_args(
+        cls,
+        retriever: BaseRetriever,
+        service_context: Optional[ServiceContext] = None,
+        node_postprocessors: Optional[List[BaseNodePostprocessor]] = None,
+        verbose: bool = False,
+        # response synthesizer args
+        response_mode: ResponseMode = ResponseMode.COMPACT,
+        text_qa_template: Optional[QuestionAnswerPrompt] = CITATION_QA_TEMPLATE,
+        refine_template: Optional[RefinePrompt] = CITATION_REFINE_TEMPLATE,
+        simple_template: Optional[SimpleInputPrompt] = None,
+        response_kwargs: Optional[Dict] = None,
+        use_async: bool = False,
+        streaming: bool = False,
+        optimizer: Optional[BaseTokenUsageOptimizer] = None,
+        # class-specific args
+        **kwargs: Any,
+    ) -> "RetrieverQueryEngine":
+        """Initialize a RetrieverQueryEngine object."
+
+        Args:
+            retriever (BaseRetriever): A retriever object.
+            service_context (Optional[ServiceContext]): A ServiceContext object.
+            node_postprocessors (Optional[List[BaseNodePostprocessor]]): A list of
+                node postprocessors.
+            verbose (bool): Whether to print out debug info.
+            response_mode (ResponseMode): A ResponseMode object.
+            text_qa_template (Optional[QuestionAnswerPrompt]): A QuestionAnswerPrompt
+                object.
+            refine_template (Optional[RefinePrompt]): A RefinePrompt object.
+            simple_template (Optional[SimpleInputPrompt]): A SimpleInputPrompt object.
+            response_kwargs (Optional[Dict]): A dict of response kwargs.
+            use_async (bool): Whether to use async.
+            streaming (bool): Whether to use streaming.
+            optimizer (Optional[BaseTokenUsageOptimizer]): A BaseTokenUsageOptimizer
+                object.
+
+        """
+        response_synthesizer = ResponseSynthesizer.from_args(
+            service_context=service_context,
+            text_qa_template=text_qa_template,
+            refine_template=refine_template,
+            simple_template=simple_template,
+            response_mode=response_mode,
+            response_kwargs=response_kwargs,
+            use_async=use_async,
+            streaming=streaming,
+            optimizer=optimizer,
+            node_postprocessors=node_postprocessors,
+            verbose=verbose,
+        )
+
+        callback_manager = (
+            service_context.callback_manager if service_context else CallbackManager([])
+        )
+
+        return cls(
+            retriever=retriever,
+            response_synthesizer=response_synthesizer,
+            callback_manager=callback_manager,
+        )
+
+    def _create_citation_nodes(self, nodes: List[NodeWithScore]) -> List[NodeWithScore]:
+        """Modify retrieved nodes to be granular sources."""
+        text_splitter = TokenTextSplitter(chunk_size=256, chunk_overlap=20)
```

Comment:
I have vague memories of the sentence splitter not working well in a few github issues/discord. But let me try here too (although there will still be some incomplete sentences since the data is pre-chunked by the default token splitter).

Reply:
seems to work fine, lets keep sentence splitter as default then (user can customize this anyways)